holoviews/streams.py

```diff
@@ -71,8 +72,10 @@ class Stream(param.Parameterized):
     respectively.
     """
 
-    # Mapping from a source id to a list of streams
-    registry = defaultdict(list)
+    # Mapping from a source to a list of streams
+    # WeakKeyDictionary to allow garbage collection
+    # of unreferenced sources
+    registry = weakref.WeakKeyDictionary()
```

Comment:
Afaik, none of the object types that we support as sources (elements, DynamicMaps, overlays etc.) implement any custom ``__hash__`` method which would interfere with them being used as keys. The reason I used the id originally was to avoid references to the source sticking around, which was actually counter-productive it turns out.